## Câu 1. Phân rã ma trận 
1. Ta có: 
$
A=
\begin{bmatrix}
1 & 0 & 1\\
0 & 1 & 1\\
1 & 0 & 1\\
0 & 1 & 1
\end{bmatrix}
$ = $
\begin{bmatrix}
1 & 0 & 1\\
0 & 1 & 1\\
0 & 0 & 0\\
0 & 0 & 0
\end{bmatrix}$. Do đó ta có $rank(A)=2 < 4$ nên từ đây suy ra A phụ thuộc tuyến tính. 
2. Ta có: $A^T = \begin{bmatrix}
1 & 0 & 1 & 0\\
0 & 1 & 0 & 1\\
1 & 1 & 1 & 1\\ \end{bmatrix}$ nên ta có : $A^T A = \begin{bmatrix}
1 & 0 & 1 & 0\\
0 & 1 & 0 & 1\\
1 & 1 & 1 & 1\\ \end{bmatrix} \begin{bmatrix}
1 & 0 & 1\\
0 & 1 & 1\\
1 & 0 & 1\\
0 & 1 & 1
\end{bmatrix}$ = $\begin{bmatrix} 2 & 0 & 2 \\ 0 & 2 & 2 \\ 2 & 2 & 4 \end{bmatrix}$
3. Xét phương trình $\det(A^T A - \lambda I) = 0 \Leftrightarrow \begin{bmatrix} 2-\lambda & 0 & 2 \\ 0 & 2-\lambda & 2 \\ 2 & 2 & 4-\lambda \end{bmatrix} = 0 \Leftrightarrow -\lambda^3 + 8 \lambda^2-12\lambda=0 \Leftrightarrow \lambda = 0$ hay $\lambda = 2$ hay $\lambda = 6$

4. Do đó ta có: $\lambda_1= 6, \lambda_2 = 2, \lambda_3 = 0$
5. Ta có: $\sigma_1 = \sqrt{6}, \sigma_2 = \sqrt{2}, \sigma_3 = 0$
6. $$\sigma_1=\sqrt{\lambda_1}=\sqrt6,\qquad \sigma_2=\sqrt{\lambda_2}=\sqrt2,\qquad \sigma_3=\sqrt{\lambda_3}=0$$

$$v_1=\frac{1}{\sqrt6}\begin{bmatrix}1\\1\\2\end{bmatrix},\qquad v_2=\frac{1}{\sqrt2}\begin{bmatrix}1\\-1\\0\end{bmatrix},\qquad v_3=\frac{1}{\sqrt3}\begin{bmatrix}1\\1\\-1\end{bmatrix}$$

 **7. Tìm eigenvector cho $\lambda_1=6$:** giải $(A^\top A-6I)v=0$

$$\begin{bmatrix}-4&0&2\\0&-4&2\\2&2&-2\end{bmatrix}v=0 \;\Rightarrow\; \begin{cases}-4v_1+2v_3=0\\-4v_2+2v_3=0\end{cases}\Rightarrow v_3=2v_1=2v_2\Rightarrow v_1=v_2$$

Chọn $v_1=v_2=1\Rightarrow v_3=2$: vector $(1,1,2)$, $\|(1,1,2)\|=\sqrt6$

$$v_1=\frac{1}{\sqrt6}(1,1,2)^\top$$

**Tìm eigenvector cho $\lambda_2=2$:** giải $(A^\top A-2I)v=0$

$$\begin{bmatrix}0&0&2\\0&0&2\\2&2&2\end{bmatrix}v=0 \;\Rightarrow\; v_3=0,\quad v_1+v_2=0\Rightarrow v_2=-v_1$$

Chọn $v_1=1,v_2=-1,v_3=0$: $\|(1,-1,0)\|=\sqrt2$

$$v_2=\frac{1}{\sqrt2}(1,-1,0)^\top$$

**Tìm eigenvector cho $\lambda_3=0$:** giải $(A^\top A)v=0$

$$\begin{bmatrix}2&0&2\\0&2&2\\2&2&4\end{bmatrix}v=0 \;\Rightarrow\; v_1=-v_3,\quad v_2=-v_3$$

Chọn $v_3=-1\Rightarrow v_1=v_2=1$: $\|(1,1,-1)\|=\sqrt3$

$$v_3=\frac{1}{\sqrt3}(1,1,-1)^\top$$

**Tìm $u_i = \dfrac{Av_i}{\sigma_i}$ với $\sigma_i \ge 0$**

**$u_1$:**

$$Av_1=A\cdot\frac{1}{\sqrt6}\begin{bmatrix}1\\1\\2\end{bmatrix}=\frac{1}{\sqrt6}\begin{bmatrix}1+0+2\\0+1+2\\1+0+2\\0+1+2\end{bmatrix}=\frac{1}{\sqrt6}\begin{bmatrix}3\\3\\3\\3\end{bmatrix}$$

$$u_1=\frac{Av_1}{\sigma_1}=\frac{1}{\sqrt6\cdot\sqrt6}\begin{bmatrix}3\\3\\3\\3\end{bmatrix}=\frac{1}{6}\begin{bmatrix}3\\3\\3\\3\end{bmatrix}=\frac12\begin{bmatrix}1\\1\\1\\1\end{bmatrix}$$

**$u_2$:**

$$Av_2=A\cdot\frac{1}{\sqrt2}\begin{bmatrix}1\\-1\\0\end{bmatrix}=\frac{1}{\sqrt2}\begin{bmatrix}1\\-1\\1\\-1\end{bmatrix}$$

$$u_2=\frac{Av_2}{\sigma_2}=\frac{1}{\sqrt2\cdot\sqrt2}\begin{bmatrix}1\\-1\\1\\-1\end{bmatrix}=\frac12\begin{bmatrix}1\\-1\\1\\-1\end{bmatrix}$$

**8. Bổ sung vector còn thiếu — Full SVD** 

Cần thêm $u_3,u_4\in\mathbb{R}^4$ trực chuẩn với $u_1,u_2$ và với nhau. Gọi $u = (a,b,c,d)$ là các vector trực chuẩn với $u_1$ và $u_2$. 

Khi đó ta có: $\begin{cases} u \bot u_1 \\ u \bot u_2 \end{cases} \Leftrightarrow \begin{cases} a + b + c + d = 0 \\ a-b+c-d=0  \end{cases} \Leftrightarrow \begin{cases} 2a+2c=0 \\ 2b+2d=0 \end{cases} \Leftrightarrow \begin{cases} c=-a \\ d=-b\end{cases}$.

Từ đây suy ra: $u=(a,b,-a,-b) = a(1,0,-1,0) + b(0,1,0,-1)$. 

Do đó ta có $u$ chính là tổ hợp tuyến tính của $u_3$ và $u_4$, chính là 2 vector cần bổ sung vào. Không mất tính tổng quát, ta chọn $u_3= (1,0,-1,0)$ và $u_4=(0,1,0,-1)$. 

Chuẩn hóa: $u_3=\dfrac{1}{\sqrt2}(1,0,-1,0)^\top$

Chuẩn hóa: $u_4=\dfrac{1}{\sqrt2}(0,1,0,-1)^\top$

**Full SVD $A=U\Sigma V^\top$:**

$$U=\begin{bmatrix}\frac12&\frac12&\frac{1}{\sqrt2}&0\\\frac12&-\frac12&0&\frac{1}{\sqrt2}\\\frac12&\frac12&-\frac{1}{\sqrt2}&0\\\frac12&-\frac12&0&-\frac{1}{\sqrt2}\end{bmatrix}\in\mathbb{R}^{4\times4},\qquad \Sigma=\begin{bmatrix}\sqrt6&0&0\\0&\sqrt2&0\\0&0&0\\0&0&0\end{bmatrix}\in\mathbb{R}^{4\times3}$$

$$V^\top=\begin{bmatrix}\frac{1}{\sqrt6}&\frac{1}{\sqrt6}&\frac{2}{\sqrt6}\\\frac{1}{\sqrt2}&-\frac{1}{\sqrt2}&0\\\frac{1}{\sqrt3}&\frac{1}{\sqrt3}&-\frac{1}{\sqrt3}\end{bmatrix}\in\mathbb{R}^{3\times3}$$

9, 10. Ta có: $A= U_r \Sigma_r V_r^T$ với $r = rank(A) = 2$. Từ đây suy ra:

$$U_r=\begin{bmatrix}\frac12&\frac12\\\frac12&-\frac12\\\frac12&\frac12\\\frac12&-\frac12\end{bmatrix}\in\mathbb{R}^{4\times2},\qquad \Sigma_r=\begin{bmatrix}\sqrt6&0\\0&\sqrt2\end{bmatrix}\in\mathbb{R}^{2\times2},\qquad V_r^\top=\begin{bmatrix}\frac{1}{\sqrt6}&\frac{1}{\sqrt6}&\frac{2}{\sqrt6}\\\frac{1}{\sqrt2}&-\frac{1}{\sqrt2}&0\end{bmatrix}\in\mathbb{R}^{2\times3}$$

Do đó: $$A = \begin{bmatrix}\frac12&\frac12\\\frac12&-\frac12\\\frac12&\frac12\\\frac12&-\frac12\end{bmatrix} \begin{bmatrix}\sqrt6&0\\0&\sqrt2\end{bmatrix}\begin{bmatrix}\frac{1}{\sqrt6}&\frac{1}{\sqrt6}&\frac{2}{\sqrt6}\\\frac{1}{\sqrt2}&-\frac{1}{\sqrt2}&0\end{bmatrix}$$

11. Ta có: $U^\top U = \begin{bmatrix}\frac12&\frac12&\frac{1}{2}&\frac12\\\frac12&-\frac12&\frac{1}{2}&-\frac{1}{2}\\\frac{1}{\sqrt{2}}&0&-\frac{1}{\sqrt2}&0\\0&\frac{1}{\sqrt{2}}&0&-\frac{1}{\sqrt2}\end{bmatrix} \begin{bmatrix}\frac12&\frac12&\frac{1}{\sqrt2}&0\\\frac12&-\frac12&0&\frac{1}{\sqrt2}\\\frac12&\frac12&-\frac{1}{\sqrt2}&0\\\frac12&-\frac12&0&-\frac{1}{\sqrt2}\end{bmatrix} = \begin{bmatrix} 1 & 0 & 0 & 0 \\ 0 & 1 & 0 & 0 \\ 0 & 0 & 1 & 0 \\ 0 & 0 & 0 & 1 \end{bmatrix} = I$ và

$$V^\top V = \begin{bmatrix}\frac{1}{\sqrt6}&\frac{1}{\sqrt6}&\frac{2}{\sqrt6}\\\frac{1}{\sqrt2}&-\frac{1}{\sqrt2}&0\\\frac{1}{\sqrt3}&\frac{1}{\sqrt3}&-\frac{1}{\sqrt3}\end{bmatrix}\begin{bmatrix}\frac{1}{\sqrt6}&\frac{1}{\sqrt2}&\frac{1}{\sqrt3}\\\frac{1}{\sqrt6}&-\frac{1}{\sqrt2}&\frac{1}{\sqrt3}\\\frac{2}{\sqrt6}&0&-\frac{1}{\sqrt3}\end{bmatrix} = \begin{bmatrix} 1 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 1 \end{bmatrix} = I$$
và 
$$U\Sigma V^\top = \begin{bmatrix}\frac12&\frac12&\frac{1}{\sqrt2}&0\\\frac12&-\frac12&0&\frac{1}{\sqrt2}\\\frac12&\frac12&-\frac{1}{\sqrt2}&0\\\frac12&-\frac12&0&-\frac{1}{\sqrt2}\end{bmatrix}\begin{bmatrix}\sqrt6&0&0\\0&\sqrt2&0\\0&0&0\\0&0&0\end{bmatrix}\begin{bmatrix}\frac{1}{\sqrt6}&\frac{1}{\sqrt6}&\frac{2}{\sqrt6}\\\frac{1}{\sqrt2}&-\frac{1}{\sqrt2}&0\\\frac{1}{\sqrt3}&\frac{1}{\sqrt3}&-\frac{1}{\sqrt3}\end{bmatrix} = \begin{bmatrix}1&0&1\\0&1&1\\1&0&1\\0&1&1\end{bmatrix} = A$$



## Kiểm tra kết quả

In [5]:
import numpy as np
np.set_printoptions(precision=4, suppress=True)
A = np.array([
    [1,0,1], 
    [0,1,1],
    [1,0,1],
    [0,1,1] ], dtype = float
)

v1 = np.array([1,1,2]) / np.sqrt(6)
v2 = np.array([1,-1,0]) / np.sqrt(2)
v3 = np.array([1,1,-1]) / np.sqrt(3)

u1 = np.array([1,1,1,1]) / 2
u2 = np.array([1,-1,1,-1]) / 2
u3 = np.array([-1,0,1,0]) / np.sqrt(2)
u4 = np.array([0,-1,0,1]) / np.sqrt(2)

U = np.column_stack([u1, u2, u3, u4])
V = np.column_stack([v1, v2, v3])
Sigma = np.array([
    [np.sqrt(6), 0, 0],
    [0, np.sqrt(2), 0],
    [0, 0, 0],
    [0, 0, 0]
])

print("U^T U =\n", U.T @ U, "\n")
print("V^T V =\n", V.T @ V, "\n")

A_reconstructed = U @ Sigma @ V.T
print("U Sigma V^T =\n", A_reconstructed, "\n")
print("Trung khop voi A?", np.allclose(A_reconstructed, A))


U^T U =
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]] 

V^T V =
 [[ 1.  0. -0.]
 [ 0.  1. -0.]
 [-0. -0.  1.]] 

U Sigma V^T =
 [[1. 0. 1.]
 [0. 1. 1.]
 [1. 0. 1.]
 [0. 1. 1.]] 

Trung khop voi A? True


In [6]:
# Doi chieu voi SVD cua numpy (dau vector co the khac, nhung cung mot duong thang)
U_np, S_np, Vt_np = np.linalg.svd(A, full_matrices=True)
print("Singular values (numpy):", S_np)
print("So sanh voi sigma tinh tay: [sqrt(6), sqrt(2), 0] =", [np.sqrt(6), np.sqrt(2), 0])

# Reduced SVD bang numpy, cat theo rank
r = np.linalg.matrix_rank(A)
Ur_np, Sr_np, Vtr_np = np.linalg.svd(A, full_matrices=False)
Ur_np, Sr_np, Vtr_np = Ur_np[:, :r], Sr_np[:r], Vtr_np[:r, :]
print("\nReduced SVD shapes -> Ur:", Ur_np.shape, "Sigma_r:", Sr_np.shape, "Vr^T:", Vtr_np.shape)
print("Tai tao tu Reduced SVD:\n", Ur_np @ np.diag(Sr_np) @ Vtr_np)
print("Trung khop voi A?", np.allclose(Ur_np @ np.diag(Sr_np) @ Vtr_np, A))

Singular values (numpy): [2.4495 1.4142 0.    ]
So sanh voi sigma tinh tay: [sqrt(6), sqrt(2), 0] = [2.449489742783178, 1.4142135623730951, 0]

Reduced SVD shapes -> Ur: (4, 2) Sigma_r: (2,) Vr^T: (2, 3)
Tai tao tu Reduced SVD:
 [[ 1.  0.  1.]
 [-0.  1.  1.]
 [ 1.  0.  1.]
 [-0.  1.  1.]]
Trung khop voi A? True


### 12. Giải thích

- **Vì sao có một singular value bằng 0?** 

Vì $\operatorname{rank}(A)=2$ trong khi $A$ có $3$ cột. Số singular value dương của một ma trận luôn đúng bằng rank của nó, nên với $A\in\mathbb{R}^{4\times3}$ có rank 2, bắt buộc phải có $3-2=1$ singular value bằng 0 ($\sigma_3=0$).
- **Singular vector tương ứng với singular value bằng 0 liên hệ thế nào với null space của $A$?** 

Ta có: $Av_3 = \sigma_3 u_3 = 0$ nên $Av_3=0$ hay nói cách khác là $v_3 \in null(A)$. 
- **Vì sao Reduced SVD vẫn tái tạo chính xác $A$?** 

Vì $A=\sum_{i=1}^{3}\sigma_iu_iv_i^\top$, và thành phần thứ 3 có $\sigma_3=0$ nên $\sigma_3u_3v_3^\top=0$ không đóng góp gì vào tổng. Do đó bỏ đi cột/hàng ứng với singular value 0 không làm mất bất kỳ thông tin nào, và $A=U_r\Sigma_rV_r^\top$ vẫn đúng tuyệt đối, không phải xấp xỉ.




# Câu 2. Chọn k cho Truncated SVD bằng tỷ lệ năng lượng

Một ma trận dữ liệu $X$ có các singular values:

$$
\sigma_1=12,\qquad
\sigma_2=5,\qquad
\sigma_3=3,\qquad
\sigma_4=1
$$

Ma trận xấp xỉ hạng $k$:

$$
X_k=\sum_{i=1}^{k}\sigma_i u_i v_i^\top
$$

## Yêu cầu

1. Xác định rank của $X$. 

Cả 4 singular value đều dương ($12,5,3,1>0$), nên khi đó ta có $\Sigma = \begin{bmatrix} 12 & 0 & 0 & 0 \\ 0 & 5 & 0 & 0 \\ 0 & 0 & 3 & 0 \\ 0 & 0 & 0 & 1 \end{bmatrix}$ là duy nhất, mà ma trận $\Sigma$ có cùng kích thước với $X$ nên $\operatorname{rank}(X)=4$.

2. Tính tổng năng lượng:

Ta có : 
$$
E_{\text{total}}=\sum_{i=1}^{4}\sigma_i^2 = 12^2 + 5^2 + 3^2 + 1^2 = 179. 
$$

### 3–4. Tính $R(k)$ và hoàn thành bảng

$$R(k)=\frac{\sum_{i=1}^k\sigma_i^2}{179}$$

- $R(1)=\dfrac{144}{179}\approx0.8045$

- $R(2)=\dfrac{144+25}{179}=\dfrac{169}{179}\approx0.9441$

- $R(3)=\dfrac{144+25+9}{179}=\dfrac{178}{179}\approx0.9944$

- $R(4)=\dfrac{179}{179}=1.0000$

| $k$ | Năng lượng giữ lại | Tỷ lệ $R(k)$ |
|---:|---:|---:|
| 1 | 144 | 0.8045 |
| 2 | 169 | 0.9441 |
| 3 | 178 | 0.9944 |
| 4 | 179 | 1.0000 |

### 5. $k$ nhỏ nhất để giữ ít nhất từng ngưỡng năng lượng

Dựa vào bảng $R(k)$ ở trên:

- **80%:** $R(1)=0.8045\ge0.80\Rightarrow k=1$
- **90%:** $R(1)=0.8045<0.90$, $R(2)=0.9441\ge0.90\Rightarrow k=2$
- **95%:** $R(2)=0.9441<0.95$, $R(3)=0.9944\ge0.95\Rightarrow k=3$
- **99%:** $R(2)=0.9441<0.99$, $R(3)=0.9944\ge0.99\Rightarrow k=3$

### 6–7. Tính $\|X-X_k\|_F$ và hoàn thành bảng

$$\|X-X_k\|_F=\sqrt{\sum_{i=k+1}^4\sigma_i^2}$$

- $k=1$: $\sqrt{5^2+3^2+1^2}=\sqrt{25+9+1}=\sqrt{35}\approx5.9161$
- $k=2$: $\sqrt{3^2+1^2}=\sqrt{9+1}=\sqrt{10}\approx3.1623$
- $k=3$: $\sqrt{1^2}=\sqrt1=1.0000$

| $k$ | $R(k)$ | $\|X-X_k\|_F$ |
|---:|---:|---:|
| 1 | 0.8045 | 5.9161 |
| 2 | 0.9441 | 3.1623 |
| 3 | 0.9944 | 1.0000 |

In [7]:
sigmas = np.array([12, 5, 3, 1], dtype=float)
sq = sigmas**2
E_total = sq.sum()
print("sigma_i^2 =", sq)
print("E_total =", E_total, "\n")

print(f"{'k':>2} | {'energy_kept':>12} | {'R(k)':>8}")
for k in range(1, 5):
    cum = sq[:k].sum()
    print(f"{k:>2} | {cum:>12.0f} | {cum/E_total:>8.4f}")

print("\nSmallest k cho tung nguong nang luong:")
for thresh in [0.80, 0.90, 0.95, 0.99]:
    for k in range(1, 5):
        cum = sq[:k].sum()
        if cum / E_total >= thresh:
            print(f"  >= {thresh*100:.0f}%  ->  k = {k}  (R(k)={cum/E_total:.4f})")
            break

print("\nFrobenius error ||X - X_k||_F:")
for k in range(1, 4):
    err = np.sqrt(sq[k:].sum())
    print(f"  k={k}: R(k)={sq[:k].sum()/E_total:.4f}, ||X-Xk||_F = {err:.4f}")

sigma_i^2 = [144.  25.   9.   1.]
E_total = 179.0 

 k |  energy_kept |     R(k)
 1 |          144 |   0.8045
 2 |          169 |   0.9441
 3 |          178 |   0.9944
 4 |          179 |   1.0000

Smallest k cho tung nguong nang luong:
  >= 80%  ->  k = 1  (R(k)=0.8045)
  >= 90%  ->  k = 2  (R(k)=0.9441)
  >= 95%  ->  k = 3  (R(k)=0.9944)
  >= 99%  ->  k = 3  (R(k)=0.9944)

Frobenius error ||X - X_k||_F:
  k=1: R(k)=0.8045, ||X-Xk||_F = 5.9161
  k=2: R(k)=0.9441, ||X-Xk||_F = 3.1623
  k=3: R(k)=0.9944, ||X-Xk||_F = 1.0000


8. Nếu $\sigma_4=1$ chủ yếu biểu diễn nhiễu, ta nên loại bỏ thành phần này khỏi xấp xỉ. Chọn **$k=3$** là hợp lý nhất vì:

- Giữ lại $R(3)=99.44\%$ năng lượng - gần như toàn bộ tín hiệu thực sự.

9. 
   - Vì sao singular values phải được sắp xếp giảm dần? 
   
   Vì ta sẽ lấy $k$ singular vaules đầu tiên ứng với ma trận cần phân rã, nghĩa là khi đó ta sẽ lấy ma trận có kích thước $k*k$ theo đường chéo chính của ma trận $\Sigma$ nên khi đó $k$ giá trị trên đường chéo chính phải là $k$ giá trị lớn nhất để đảm bảo đủ năng lượng để tái tạo ma trận cần phân rã.  
   - Giữ càng nhiều singular values có phải lúc nào cũng tốt hơn không? 

Không. Giữ nhiều hơn giúp giảm sai số tái tạo $\|X-X_k\|_F$, nhưng đồng thời: (1) giữ lại cả nhiễu — như $\sigma_4$ ở câu 8, vốn không mang tín hiệu hữu ích; (2) tăng chi phí lưu trữ/tính toán, làm giảm hiệu quả nén; (3) có thể gây overfitting nếu dùng cho các tác vụ học máy tiếp theo (ví dụ giảm chiều dữ liệu trước khi phân loại).
   - Sự đánh đổi giữa tỷ lệ năng lượng và mức độ nén dữ liệu là gì? 
   
Ta có : $$ CR = \frac{4mn}{k(m+n+1)}$$ 
$k$ càng lớn thì $R(k)$ càng cao (tái tạo càng chính xác) nhưng mức nén càng thấp (cần lưu nhiều $u_i,\sigma_i,v_i$ hơn). $k$ càng nhỏ thì nén càng mạnh (lưu ít dữ liệu hơn) nhưng $R(k)$ thấp hơn, sai số $\|X-X_k\|_F$ lớn hơn. Việc chọn $k$ luôn là bài toán cân bằng giữa độ chính xác cần thiết cho tác vụ và giới hạn tài nguyên (bộ nhớ, tốc độ tính toán).